### PR TITLE
Missing includes

### PR DESCRIPTION
**Pull request by dkeeney**
See Issue #885 

I tried making a copy of the napi_hello example and build it without using CMake.
This PR attempts to fix the problems that were found.

- cereal include files were not found
- cpp-htmlib include file was not found (would be needed by RESTapi examples.
- Global in htm/os/timer.cpp needed to be corrected.
- importFileSystem.hpp did not have the right logic for avoiding boost
- Added some sample build scripts
- Added comments to README.md
- Undefined references when compiling with GCC for debug mode (without Python).

### CODE REVIEW

**Comment by dkeeney** [comment]
I am still getting this `undefined references` error when linking to a debug version of htm.core (shared library).
(Compiling a copy of napi_hello.cpp)
```
ave@ubuntu1810:~/htm.core/examp$ ./debugBuild.sh 
/usr/bin/ld: /tmp/ccIclmkO.o: in function `void htm::SparseDistributedRepresentation::load_ar<cereal::BinaryInputArchive>(cereal::BinaryInputArchive&)':
/home/dave/htm.core/examp/../build/Debug/include/htm/types/Sdr.hpp:616: undefined reference to `htm::SparseDistributedRepresentation::initialize(std::vector<unsigned int, std::allocator<unsigned int> > const&)'
/usr/bin/ld: /tmp/ccIclmkO.o: in function `void htm::SparseDistributedRepresentation::load_ar<cereal::PortableBinaryInputArchive>(cereal::PortableBinaryInputArchive&)':
/home/dave/htm.core/examp/../build/Debug/include/htm/types/Sdr.hpp:616: undefined reference to `htm::SparseDistributedRepresentation::initialize(std::vector<unsigned int, std::allocator<unsigned int> > const&)'
/usr/bin/ld: /tmp/ccIclmkO.o: in function `void htm::SparseDistributedRepresentation::load_ar<cereal::JSONInputArchive>(cereal::JSONInputArchive&)':
/home/dave/htm.core/examp/../build/Debug/include/htm/types/Sdr.hpp:616: undefined reference to `htm::SparseDistributedRepresentation::initialize(std::vector<unsigned int, std::allocator<unsigned int> > const&)'
/usr/bin/ld: /tmp/ccIclmkO.o: in function `void htm::SparseDistributedRepresentation::load_ar<cereal::XMLInputArchive>(cereal::XMLInputArchive&)':
/home/dave/htm.core/examp/../build/Debug/include/htm/types/Sdr.hpp:616: undefined reference to `htm::SparseDistributedRepresentation::initialize(std::vector<unsigned int, std::allocator<unsigned int> > const&)'
/usr/bin/ld: /tmp/ccIclmkO.o:(.data.rel.ro._ZTVN3htm5ArrayE[_ZTVN3htm5ArrayE]+0x60): undefined reference to `htm::ArrayBase::allocateBuffer(std::vector<unsigned int, std::allocator<unsigned int> > const&)'
collect2: error: ld returned 1 exit status
```

- Release mode seems ok.
- Debug mode is ok when building under CMake.

This type of error normally has something to do with using templates that are not entirely defined at compile time (not everything in .hpp) but I cannot determine where the problem is.  

@breznak  @Zbysekz  any ideas of how to go about locating this problem?

File: README.md [hunk 1]
```diff
@@ -148,13 +152,59 @@ make -j8 install
...
+#      cd build/scripts
+#      cmake ../..
+#      make -j4 install
+#
+# Now build myapp
+# We use -std=c++17 to get <filesystem> so we can avoid using the boost library.
+# The -I gives the path to the includes needed to use with the htm.core library.
+# The -L gives the path to the shared htm.core library location at build time.
+# The LD_LIBRARY_PATH envirment variable points to the htm.core library location at runtime.
+g++ -o myapp -std=c++17  -I <path-to-repo>/build/Release/include myapp.cpp -L <path-to-repo>/build/Release/lib -lhtm_core -lpthread -ldl
```
Comment (breznak) [inline README.md:178]: this is very nice :+1: do we want to add it as a "demo" step to the CI, or Readme is enough?

File: README.md [hunk 2]
```diff
@@ -148,13 +152,59 @@ make -j8 install
...
+
+# Run myapp 
+export LD_LIBRARY_PATH=<path-to-repo>/build/Release/lib:$LD_LIBRARY_PATH
+./myapp
```
Comment (breznak) [inline README.md:182]: nit: `LD_LIBRARY_PATH=xxx ./myapp` should also work

File: README.md [hunk 3]
```diff
@@ -148,13 +152,59 @@ make -j8 install
...
+
+Here is an example of a **Debug build** of your own C++ app that links to htm.core as a shared library.
+```
+#! /bin/sh
+# Using GCC on linux ...
+# First build htm.core as debug from sources.
+#      cd <path-to-repo>
+#      mkdir -p build/scripts
+#      cd build/scripts
+#      cmake ../.. -DCMAKE_BUILD_TYPE=Debug
```
Comment (breznak) [inline README.md:193]: maybe to keep reasonable size of the Readme, we don't need this detail (?) and just say "build htm.core in Debug mode" ?

File: README.md [hunk 4]
```diff
@@ -148,13 +152,59 @@ make -j8 install
...
+#      make -j4 install
+#
+# Now build myapp
+# The -g -Og tells the compiler to build debug mode with no optimize.
+# We use -std=c++17 to get <filesystem> so we can avoid using the boost library.
+# The -I gives the path to the includes needed to use with the htm.core library.
+# The -L gives the path to the shared htm.core library location at build time.
+# The LD_LIBRARY_PATH envirment variable points to the htm.core library location at runtime.
+g++ -g -Og -o myapp -std=c++17  -I <path-to-repo>/build/Debug/include myapp.cpp -L <path-to-repo>/build/Debug/lib -lhtm_core -lpthread -ldl
```
Comment (breznak) [inline README.md:202]: ..but keep this, that's good :+1:

File: README.md [hunk 5]
```diff
@@ -148,13 +152,59 @@ make -j8 install
...
+
+# Run myapp in the debugger
+export LD_LIBRARY_PATH=<path-to-repo>/build/Debug/lib:$LD_LIBRARY_PATH
+gdb ./myapp
```
Comment (breznak) [inline README.md:206]: this is useful too!

File: src/htm/os/ImportFilesystem.hpp [hunk 6]
```diff
@@ -54,8 +54,13 @@
   #endif
 #else
   // C++11
-  #include <boost/filesystem.hpp>
-  #define USE_BOOST_FILESYSTEM 1
+  #if (defined( __GNUC__ ) && __GNUC__ >= 8) || (defined(__clang_major__) && __clang_major__ >= 7)
+      #include <filesystem>
```
Comment (breznak) [inline src/htm/os/ImportFilesystem.hpp:58]: not sure this is correct, imho users just should know to specify `-std=c++17` when compiling their app with htm.core. As you could have gcc-8.3 but use c++11/98 in your project and cannot use TS filesystem then.

File: src/examples/napi_hello/DebugBuild.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#      cmake ../.. -DCMAKE_BUILD_TYPE=Debug
+#      make -j4 install
+#
+# Now build napi_hello
+# The -g -Og tells the compiler to build debug mode with no optimize.
+# We use -std=c++17 to get <filesystem> so we can avoid using the boost library.
+# The -I gives the path to the includes needed to use the htm.core library.
+# The -L gives the path to the shared htm.core library location at build time.
+# -lhtm_core says link with libhtm_core.so
+g++ -g -Og -o napi_hello -std=c++17  -I ../../../build/Debug/include myapp.cpp -L ../../../build/Debug/lib -lhtm_core -lpthread -ldl
```
Comment (breznak) [inline src/examples/napi_hello/DebugBuild.sh:16]: in the final PR, I'm (weakly) against including (specific) build scripts. The readme should be enough, maybe also rewrite this as an example step in the CI (like we have for valgrind etc) and point to it from the readme section too.

**Review by breznak — COMMENTED** [review]
This looks nice, thank you. 
Some feedback below. 

Unfortunately, I don't see any reason for the failing Debug linking. Maybe try asking at Cereal why this use-case keeps failing? (but we can definitely merge even w/o the Debug support, although it'd be nice for sure)

**Comment by Zbysekz** [comment]
> his type of error normally has something to do with using templates that are not entirely defined at compile time (not everything in .hpp) but I cannot determine where the problem is.
> 
> @breznak @Zbysekz any ideas of how to go about locating this problem?

Yes it is because of mismatch of the "standard" library, specifically `vector`. 

In libhtm_core.so is 
`00000000004291ba T htm::SparseDistributedRepresentation::initialize(std::__debug::vector<unsigned int, std::allocator<unsigned int> > const&)`
but in include file
`htm::SparseDistributedRepresentation::initialize(std::vector<unsigned int, std::allocator<unsigned int> > const&)`

Htm core is compiled in debug, so this application std library must also be compiled with debug with compiler flag `-D_GLIBCXX_DEBUG`
Original answer [https://github.com/htm-community/htm.core/issues/880#issuecomment-684721452](https://github.com/htm-community/htm.core/issues/880#issuecomment-684721452)

File: README.md (hunk 3 above)
Comment (Zbysekz) [reply README.md:193]: I would keep this, but create new readme section for "Building from source" and put the whole section there? We could also add there "FAQ - common problems with compiling" for example

**Comment by dkeeney** [comment]
> -D_GLIBCXX_DEBUG

That was it!!!!   Thank you @Zbysekz 
It makes perfect since now that you mention it.

File: src/htm/os/ImportFilesystem.hpp (hunk 6 above)
Comment (dkeeney) [reply src/htm/os/ImportFilesystem.hpp:58]: > imho users just should know to specify -std=c++17 when compiling their app with htm.core. 

Well, I did not know that.  I had to go figure it out when it kept trying to include boost.

File: README.md (hunk 2 above)
Comment (breznak) [reply README.md:182]: > But as a normal practice it is best to keep the original LD_LIBRARY_PATH. 

sure, I meant avoiding the `export` and applying the override only for the ./app call (but that's just a nuance)

File: README.md (hunk 3 above)
Comment (breznak) [reply README.md:193]: FAQ sounds good

File: src/examples/napi_hello/napi_hello.hpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * --------------------------------------------------------------------- */
+
+#include <htm/engine/Network.hpp>
+#include <htm/algorithms/AnomalyLikelihood.hpp>
```
Comment (breznak) [inline src/examples/napi_hello/napi_hello.hpp:19]: is this file complete/needed?

File: src/examples/napi_hello/DebugBuild.sh [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+#      make -j4 install
+#
+# Now build napi_hello
+# The -g -Og tells the compiler to build debug mode with no optimize.
+# We use -std=c++17 to get <filesystem> so we can avoid using the boost library.
+# The -D_GLIBCXX_DEBUG setting says compile std:: with debug
+# The -I gives the path to the includes needed to use the htm.core library.
+# The -L gives the path to the shared htm.core library location at build time.
+# -lhtm_core says link with libhtm_core.so
+g++ -g -Og -o napi_hello -std=c++17 -D_GLIBCXX_DEBUG -I ../../../build/Debug/include myapp.cpp -L ../../../build/Debug/lib -lhtm_core -lpthread -ldl
```
Comment (breznak) [inline src/examples/napi_hello/DebugBuild.sh:17]: as I mentioned, I'd prefer removing these example scripts, as we have it in the readme. But if you want to keep them, it's ok.

**Review by breznak — APPROVED** [review]

File: src/examples/napi_hello/napi_hello.hpp (hunk 8 above)
Comment (dkeeney) [reply src/examples/napi_hello/napi_hello.hpp:19]: oops,   That was something that I tried.
I will remove it.

File: src/examples/napi_hello/DebugBuild.sh (hunk 9 above)
Comment (dkeeney) [reply src/examples/napi_hello/DebugBuild.sh:17]: Let's keep the sample sh scripts for now.  It is for new C++ programmers.

**Comment by dkeeney** [comment]
hmmm, I guess I have to create a new PR to remove that .hpp file.

**Pull request merged**